pkg/metrics/collector/netlink_metrics.go

```diff
@@ -15,85 +15,287 @@ package collector
+}
+
+// Functionality for multithreaded map usage
+func safeIPWrite(key string, value podDef) {
+	mapMux.Lock()
+	defer mapMux.Unlock()
+	ipMap[key] = value
+}
+
+func safeIPDelete(key string) {
+	mapMux.Lock()
+	defer mapMux.Unlock()
+	delete(ipMap, key)
+}
+
+func safeIPRead(key string) (podDef, bool) {
+	mapMux.RLock()
+	defer mapMux.RUnlock()
+	podDef, ok := ipMap[key]
+	return podDef, ok
+}
+
+// Functionality for pod watching
+func createPodWatch() error {
+	// creates the in-cluster config
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		return err
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	watchlist := cache.NewListWatchFromClient(
+		clientset.CoreV1().RESTClient(),
+		"pods",
+		v1.NamespaceAll,
+		fields.OneTermEqualSelector("spec.nodeName", nodeName),
+	)
+	_, controller := cache.NewInformer(
+		watchlist,
+		&v1.Pod{},
+		0,
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    onPodAdd,
+			UpdateFunc: onPodUpdate,
+			DeleteFunc: onPodDelete,
+		},
+	)
+
+	stopper := make(chan struct{})
+
+	go controller.Run(stopper)
+
+	return nil
+}
+
+func podAllow(pod *v1.Pod) bool {
+	// Pod not assigned.
+	if pod.Status.PodIP == "" {
+		return false
+	}
+	// Pod running on host network.
+	if pod.Status.PodIP == pod.Status.HostIP {
+		return false
+	}
+	return true
+}
+
+func onPodAdd(podObj interface{}) {
+	pod := podObj.(*v1.Pod)
+	if podAllow(pod) {
+		safeIPWrite(pod.Status.PodIP, podDef{name: pod.ObjectMeta.Name, namespace: pod.ObjectMeta.Namespace})
+	}
+}
+
+func onPodUpdate(oldPod, newPod interface{}) {
+	prevPod := oldPod.(*v1.Pod)
+	pod := newPod.(*v1.Pod)
+	if prevPod.Status.PodIP == pod.Status.PodIP {
+		return
+	}
+	if podAllow(prevPod) {
+		safeIPDelete(prevPod.Status.PodIP)
+	}
+	if podAllow(pod) {
+		safeIPWrite(pod.Status.PodIP, podDef{name: pod.ObjectMeta.Name, namespace: pod.ObjectMeta.Namespace})
+	}
+}
+
+func onPodDelete(podObj interface{}) {
+	pod := podObj.(*v1.Pod)
+	if podAllow(pod) {
+		safeIPDelete(pod.Status.PodIP)
+	}
+}
+
+func updatePodStats(stats netlinkStats, tcpInfo *tcp.LinuxTCPInfo) netlinkStats {
+	stats.tcpConnections++
+	if tcpInfo == nil {
+		return stats
+	}
+	stats.retransmits += uint64(tcpInfo.Retrans)
+	return stats
+}
+
+func createStatMap(snapshots []*parser.Snapshot) map[podDef]netlinkStats {
+	statMap := make(map[podDef]netlinkStats)
+	for _, snapshot := range snapshots {
+		if snapshot.SockInfo == nil {
+			continue
+		}
+		if pod, ok := safeIPRead(snapshot.SockInfo.SrcIP); ok {
+			statMap[pod] = updatePodStats(statMap[pod], snapshot.TCPInfo)
+		}
+
+		if pod, ok := safeIPRead(snapshot.SockInfo.DstIP); ok {
+			statMap[pod] = updatePodStats(statMap[pod], snapshot.TCPInfo)
+		}
+	}
+	return statMap
+}
+
+func NewNetlinkCollector() (Collector, error) {
+	return &netlinkCollector{}, nil
+}
+
+func getPids() ([]int, error) {
+	files, err := ioutil.ReadDir("/proc")
```

Comment:
Humm, I honestly don't know - would the actual `/proc` folder from the host be shared with the netd pod? Do we need to add any escalated privilege for this?

Reply:
Yeah, unfortunately we need to have the namespaces of the other containers on the node which requires access to their proc folders. This requires hostPID to be set to true in the netd pod spec. Without it we can only read our namespace since containerd/docker don't expose the namespaces of other containers in the cluster. Let me know what you think about this.